sys/filedesc.c

```diff
@@ -26,18 +26,51 @@ static void fd_mark_unused(fdtab_t *fdt, int fd) {
   bit_clear(fdt->fdt_map, fd);
 }
 
+/* Grows given file descriptor table to contain new_size file descriptors
+ * (up to MAXFILES)
+ */
+
+static void fd_growtable(fdtab_t *fdt, size_t new_size) {
+  assert(fdt->fdt_nfiles < new_size && new_size <= MAXFILES);
+
+  file_t **old_fdt_files = fdt->fdt_files;
+  bitstr_t *old_fdt_map = fdt->fdt_map;
+
+  file_t **new_fdt_files =
+    kmalloc(fd_pool, sizeof(file_t *) * new_size, M_ZERO);
+  bitstr_t *new_fdt_map = kmalloc(fd_pool, bitstr_size(new_size), M_ZERO);
+  memset(new_fdt_map, 0, bitstr_size(fdt->fdt_nfiles * 2));
+
+  memcpy(new_fdt_files, old_fdt_files, sizeof(file_t *) * fdt->fdt_nfiles);
+  memcpy(new_fdt_map, old_fdt_map, bitstr_size(fdt->fdt_nfiles));
+  kfree(fd_pool, old_fdt_files);
+  kfree(fd_pool, old_fdt_map);
+
+  fdt->fdt_files = new_fdt_files;
+  fdt->fdt_map = new_fdt_map;
+  fdt->fdt_nfiles *= 2;
```

Comment:
This assumes that `new_size == 2 * fdt->fdt_nfiles` . This happens to be true, but you probably want to assign `new_size` to `nfiles` here.